dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AliasableResourceInterface.java

```diff
@@ -0,0 +1,66 @@
+/*
+ *    Copyright 2019 OICR
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+package io.dockstore.webservice.resources;
+
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeSet;
+
+import com.google.common.collect.Sets;
+import io.dockstore.webservice.core.Alias;
+import io.dockstore.webservice.core.Aliasable;
+import io.dockstore.webservice.core.Entry;
+import io.dockstore.webservice.core.User;
+import io.dockstore.webservice.helpers.ElasticManager;
+import io.dockstore.webservice.helpers.ElasticMode;
+
+public interface AliasableResourceInterface<T extends Aliasable> {
+
+    Optional<ElasticManager> getElasticManager();
+
+    /**
+     * Get a resource with id and only return it if user has rights to see/change it
+     * @param user
+     * @param id
+     * @return
+     */
+    T getAndCheckResource(User user, Long id);
+
+    /**
+     * Get aliases, only works on public entries for now
+     * @param alias
+     * @return
+     */
+    T getAndCheckResourceByAlias(String alias);
+
+    default T updateAliases(User user, Long id, String aliases, String emptyBody) {
+        T c = getAndCheckResource(user, id);
+        // compute differences
+        Set<String> oldAliases = c.getAliases().keySet();
+        Set<String> newAliases = Sets.newHashSet(Arrays.stream(aliases.split(",")).map(String::trim).toArray(String[]::new));
+        Set<String> aliasesToAdd = Sets.difference(newAliases, oldAliases);
+        Set<String> aliasesToRemove = new TreeSet<>(Sets.difference(oldAliases, newAliases));
+        // add new ones and remove old ones while retaining the old entries and their order
+        aliasesToAdd.forEach(alias -> c.getAliases().put(alias, new Alias()));
```

Comment:
I may be misunderstanding this, but doesn't each `c.getAliases().put()` end up writing to the database?

If yes, then isn't there the possibility that one of them could fail because the alias is not unique?

If so, and any aliases were updated before the failure case, then elastic search index would not get updated.

Reply:
`aliasesToAdd = Sets.difference(newAliases, oldAliases)` should mean that aliasesToAdd only contains aliases that don't already exist. @UnitOfWork should ensure that this is happening in a transaction.

Will add a test to double-check

Reply:
Yup, although on second thought. 
We may want to add an endpoint that simply adds aliases without allowing one to delete one, but I think this demonstrates in general what we want.

Reply:
I was thinking more of the case where another user got one of the aliases before this code executed -- aliases are unique across of all Dockstore, aren't they?

Regarding `@unitofwork`, if I understand correctly, that would lead to a different problem -- the ES index will have been updated, but then if there is an error saving the data and a rollback gets triggered, the ES index would reflect the pre rolled back values.

These are edge cases though, not crucial to fix.

Reply:
Oh, I see. 
Yes, the elastic search part won't rollback properly in the case that another user used the same alias on a different tool. The database transaction should.